File: algorithms/cpp/dataStreamasDisjointIntervals/dataStreamasDisjointIntervals.cpp
// Source: https://leetcode.com/problems/data-stream-as-disjoint-intervals/
// Author: Miao Zhang
// Date:   2021-02-04

class SummaryRanges {
public:
    /** Initialize your data structure here. */
    SummaryRanges() {
        
    }
    
    void addNum(int val) {
        vector<int> new_interval{val, val};
        int i;
        int n = intervals.size();
        int overlap = 0;
        for (i = 0; i < n; i++) {
            if (new_interval[1] + 1 < intervals[i][0]) break;
            if (new_interval[0] <= intervals[i][1] + 1) {
                new_interval[0] = min(new_interval[0], intervals[i][0]);
                new_interval[1] = max(new_interval[1], intervals[i][1]);
                overlap++;
            }
        }
        if (overlap > 0) {
            intervals.erase(intervals.begin() + i - overlap, intervals.begin() + i);
        }
        intervals.insert(intervals.begin() + i - overlap, new_interval);
    }
    
    vector<vector<int>> getIntervals() {
        return intervals;
    }

private:
    vector<vector<int>> intervals;
};

/**
 * Your SummaryRanges object will be instantiated and called as such:
 * SummaryRanges* obj = new SummaryRanges();
 * obj->addNum(val);
 * vector<vector<int>> param_2 = obj->getIntervals();
 */

File: algorithms/python/dataStreamasDisjointIntervals/dataStreamasDisjointIntervals.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/data-stream-as-disjoint-intervals/
# Author: Miao Zhang
# Date:   2021-02-04

class SummaryRanges:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.intervals = []

    def addNum(self, val: int) -> None:
        new_interval = [val, val]
        n = len(self.intervals)
        overlap = 0
        i = 0
        while i < n:
            if new_interval[1] + 1 < self.intervals[i][0]:
                break
            if new_interval[0] <= self.intervals[i][1] + 1:
                new_interval[0] = min(new_interval[0], self.intervals[i][0])
                new_interval[1] = max(new_interval[1], self.intervals[i][1])
                overlap += 1
            i += 1
        del self.intervals[i - overlap: i]
        self.intervals.insert(i - overlap, new_interval)
        
    def getIntervals(self) -> List[List[int]]:
        return self.intervals

# Your SummaryRanges object will be instantiated and called as such:
# obj = SummaryRanges()
# obj.addNum(val)
# param_2 = obj.getIntervals()
